### ISSUE TITLE
Consider not checking timestamp every time for every helper call

### ISSUE THREAD

**Issue by justin808** [rails/webpacker#662]
In our app https://www.friendsandguests.com, we currently have roughly 8,000 client side files and growing.

We've got thousands of tests. 

We'll also be including multiple bundles on many pages.

Could we consider changing the following code so that it only runs once at the initialization of tests? This is the way we currently have the code setup in React on Rails, [per these docs](https://github.com/shakacode/react_on_rails/blob/master/docs/additional-reading/rspec-configuration.md).

https://github.com/rails/webpacker/blob/master/lib/webpacker/compiler.rb#L38

```rb
    def source_last_changed_at
      files = Dir[*default_watched_paths, *watched_paths].reject { |f| File.directory?(f) }
      files.map { |f| File.mtime(f).utc.to_i }.max.to_s
    end
```

Since we already have a `cache_manifest` setting, we could also have a `cache_source_last_changed_at` setting. 

> Set cache_source_last_changed_at to true if you would like this value computed only once for compilation, such as during a test run.

**Comment by gauravtiwari** [comment]
@justin808 How much overhead this is adding? Did you benchmark it?

**Comment by gauravtiwari** [comment]
Ohh this is for source, I thought manifest. You can turn off the compiler `compile: false` and use watcher or dev server no? Sorry, don't fully understand the problem :)

**Comment by justin808** [comment]
I didn't benchmark this, but this is going to be a huge number of files to
make a system call to.

**Comment by gauravtiwari** [comment]
@justin808 Yepp but that's the idea actually with on-demand lazy compilation - suitable only for one off builds (large projects), like for demo or for small projects where there is just few packs. 

For large number of packs and if developing it's advised to either use dev-server or watcher since they will be significantly faster for that use case.

**Comment by justin808** [comment]
@gauravtiwari My use cases are:

1. No CI setup for ensuring test compilation. 
2. Developers run the watch or dev-server for development files. Run tests at cmd line.

In both #1 and #2, this is how we do it for LARGE projects.

Keep in mind, we have different test bundles than dev bundles b/c of extra logging for dev.

**Comment by gauravtiwari** [comment]
@justin808 Right so I just benchmarked it by including whole `node_modules` into the `watched_paths`. The node_modules folder has 33000 files. Takes about 3s to compute `source_last_changed_at` 

![screen shot 2017-08-18 at 19 04 46](https://user-images.githubusercontent.com/771039/29471546-47335386-8448-11e7-9ce4-f2c7b58801b6.png)

**Comment by gauravtiwari** [comment]
But I guess in CI if you have that many files and computing this for all requests would add up a lot more time.

**Comment by gauravtiwari** [comment]
cc// @dhh

**Comment by gauravtiwari** [comment]
Perhaps we can use same `cache_manifest` option to turn off compile if it's set to true but it can be confusing or may be not 

https://github.com/rails/webpacker/blob/master/lib/webpacker/manifest.rb#L31
```rb
  def compiling?
    config.compile? && !dev_server.running? && !config.cache_manifest?
  end
```

**Comment by justin808** [comment]
@gauravtiwari Excellent point! If the manifest is cached then the generated webpack files should be cached as well.

To be honest, what I really need is the following when cache_manifest is set to true:

1. caching is lazy. When Ruby loads the classes, the manifest may not exist yet, and @data.empty? == true
2. if @data.empty?, then the manifest will try to be reloaded, and it may still not exist. Then the compilation can run if compile is true.
3. So long as the compile can call a customized command, then React on Rails can use this. The key could be called something like `compile_command` and be env specific. If set to `default` then the default webpacker one is used.

Another issue is that checking if the `dev_server.running?` seems environment specific. In other words, I'm using the dev server for dev manual testing, and I want to run my tests in a new shell.

**Comment by dhh** [comment]
I don't understand why React on Rails needs to rely on the compiler command used by Webpacker? As long as you set `compile: false`, then you'll have responsibility of compiling the files through the dev server or the watcher. How you run those isn't specified by Webpacker.

And if you're compiling your stuff via the watcher or the dev server, you'll be continuously generating a new manifest. Which means that you don't want to be caching that.

dev_server.running? depends on the dev_server options being set in webpacker.yml. They're only set for the development environment, so that method will return false for test.

**Comment by dhh** [comment]
For your testing case, why not just compile all the files prior to the run, then call Webpacker.bootstrap, and run with compile: false and cache_manifest: true? Basically run it like you would production.

**Comment by justin808** [comment]
@dhh I was considering having "React on Rails" to depend on the lazy compilation feature of Webpacker for tests. 

React on Rails has long had a test helper that would lazily check for out of date client files at most once for a test run, and only from the type of test that would hit the view helpers, such as a feature spec, as documented [here](https://github.com/shakacode/react_on_rails/blob/master/docs/additional-reading/rspec-configuration.md).

**Comment by dhh** [comment]
I think it's probably either or. Either you depend on the lazy compilation, and then there's no caching. Or you cache up front, and turn off lazy compilation. Whatever fits the needs of React on Rails better.
